### ISSUE TITLE
Usage/meaning for dependencies

### ISSUE THREAD

**Issue by dionmcm** [IHTSDO/termserver-syndication#4]
What I think needs clarifying in this document is when these dependencies are stated and what that means.

### Edition versus Extension RF2 packages
I think we've got 2 different types of RF2 release, and the implication of stating dependencies is different for each
- edition, containing at least all mandatory dependencies premixed into a stand alone set of files
- extension, containing just the additions to some base edition and can't stand on its own without that base edition content

The current [NCTS specification SNOMED CT content type](https://www.healthterminologies.gov.au/specs/v3/content-types/snomed-ct/) isn't explicit about which of these two things is being described, although it does use SNOMED CT Edition URIs which is uses to identify these RF2 packages and all of the RF2 packages released on the NCTS are edition packages, not extension packages.

So I think the first issue to resolve is if we're going to have both edition and extension RF2 packages in a feed, how do we tell which is which? I think to do that we'd need a parallel set of categories to [the ones in the NCTS specification](https://www.healthterminologies.gov.au/specs/v3/conformant-server-apps/syndication-api/syndication-feed/). Perhaps there are more elegant solutions, but I can't think of one - the URI specification only has URIs for editions, not extensions.

Existing | NCTS Content Type | Distribution Format | Atom Category term value | Atom Category label value
-- | -- | -- | -- | --
Existing | SNOMED CT | Edition Delta | SCT_RF2_DELTA | SNOMED CT Edition RF2 Delta
Existing | SNOMED CT | Edition Snapshot | SCT_RF2_SNAPSHOT | SNOMED CT Edition RF2 Snapshot
Existing | SNOMED CT | Edition Full | SCT_RF2_FULL | SNOMED CT Edition RF2 Full
Existing | SNOMED CT | Edition All | SCT_RF2_ALL | SNOMED CT Edition RF2 All
New | SNOMED CT | Extension Delta | SCT_EXT_RF2_DELTA | SNOMED CT Extension RF2 Delta
New | SNOMED CT | Extension Snapshot | SCT_EXT_RF2_SNAPSHOT | SNOMED CT Extension RF2 Snapshot
New | SNOMED CT | Extension Full | SCT_EXT_RF2_FULL | SNOMED CT Extension RF2 Full
New | SNOMED CT | Extension All | SCT_EXT_RF2_ALL | SNOMED CT Extension RF2 All

### What do the dependencies mean?
Then there's the question of what the presence/absences of the set of dependencies might mean for the entry, which might differ depending on whether the entry represents an edition or an extension.
```
<sct:packageDependency>
     <sct:editionDependency></sct:editionDependency>
     <sct:derivativeDependency></sct:derivativeDependency>
</sct:packageDependency>
```

Dependencies using this extension could mean either
1. external content not included in the package but necessary to use the content - a true dependency
2. metadata advertising external dependencies that have already been mixed into the package

Number 1 doesn't really make sense for an edition package, by definition they contain all their dependencies premixed. However this is essential for an extension package.

Number 2 could really apply to either edition or extension packages. That said, typically extension packages are likely to be minimalist without extras mixed in, while editions are more likely to contain "extras" outside strictly necessary dependencies.

I think the main thing is we need to be clear about which of these two things a set of package dependencies means when it appears in an entry.

Perhaps the simplest approach would be to say sct:packageDependency means number 1 above, and therefore editions won't express any dependencies because they always include all their required content. If we want to do number 2, advertise the mandatory/optional edition/derivative versions that have been included in the package we create a separate Atom extension to representation that.

**Comment by dionmcm** [comment]
Actually with this in mind, it might be good to include and example of an extension RF2 package entry as well as an edition RF2 package entry in the example feed. I'm happy to do the work if we can agree how we're going to identify these things.

Maybe adding some commentary to the example feed to highlight things in the example would be good too.

**Comment by kaicode** [comment]
It's useful to know that the existing feed only contains edition packages.

I believe the intention is that package dependencies are only listed when they required, for loading an extension package. An edition package would never list any dependencies because for the purpose of consuming the package there are no dependencies. 

With that in mind there may be no need for new Atom Category term values, because extensions could clearly be identified by those entries that have a `sct:packageDependency` node with at least one child? I originally thought that including this node even when it is empty would be a good way to explicitly declare that there are no dependencies, but if the original feed doesn't have any extensions maybe this is not necessary.

The URI standard does not have URIs for extensions because an extension is only a package format. Logically an extension plus its dependencies is an edition.

**Comment by dionmcm** [comment]
So

- A `sct:packageDependency` element is only included where there are required dependencies that are not present in the package
- edition packages (International or otherwise) will therefore not contain a `sct:packageDependency` element
- extension packages will always contain a `sct:packageDependency` element
- edition vs extension packages can be identified by the absence or presence of `sct:packageDependency` elements.

That works for me - I'll create a PR with some suggested changes to clarify this.

**Comment by dionmcm** [comment]
Actually @kaicode one other clarification - `<sct:editionDependency></sct:editionDependency>` versus `<sct:derivativeDependency></sct:derivativeDependency>`. Aren't these both just SNOMED CT URIs?

What is the distinction here? Is it that `<sct:derivativeDependency></sct:derivativeDependency>` will always be an extension package?

Why do we need two separate lists of dependency URIs?

**Comment by rorydavidson** [comment]
@dionmcm Our thinking was to call it out specifically, considering there are several derivatives that are packaged/released on their own. It's a fair point that they are 'just' another extension package, it was just to be explicit. Do you see any issues with this?

**Comment by dionmcm** [comment]
Not particularly @rorydavidson, I don't have a strong preference either way. I was just trying to picture what the difference is so that we can be explicit in the documentation and not have someone ask the question later ;-)

So in the example 
```
    <entry>
        <title>SNOMED CT-International Spanish Edition 2022-10-31 (RF2 SNAPSHOT)</title>
        <link rel="alternate" type="application/zip" href="https://mlds.ihtsdotools.org/api/releasePackages/167/releaseVersions/849892/releaseFiles/849900/download" length="266118090" ncts:sha256Hash="902ec61f6ada329bfd6bc487fe7adab0" />
        <category term="SCT_RF2_SNAPSHOT" label="SNOMED CT RF2 Snapshot" scheme="http://ns.electronichealth.net.au/ncts/syndication/asf/scheme/1.0.0" />
        <author>
            <name>SNOMED International</name>
            <uri>http://www.snomed.org</uri>
            <email>info@snomed.org</email>
        </author>
        <id>urn:uuid:f9df5b1f-bbd3-4b25-ac7c-13a2cfaa3184</id>
        <rights>Copyright International Health Terminology Standards Development Organisation 2002-2022.  All rights reserved.  SNOMED CT® was originally created by the College of American Pathologists.  "SNOMED" and "SNOMED CT" are registered trademarks of International Health Terminology Standards Development Organisation, trading as SNOMED International.</rights>
        <updated>2022-08-01T06:00:00Z</updated>
        <published>2022-08-04T06:00:00Z</published>
        <summary></summary>
        <ncts:contentItemIdentifier>http://snomed.info/sct/450829007</ncts:contentItemIdentifier>
        <ncts:contentItemVersion>http://snomed.info/sct/450829007/version/20221031</ncts:contentItemVersion>
        <sct:packageDependency>
            <sct:editionDependency>http://snomed.info/sct/900000000000207008/version/20220731</sct:editionDependency>
            <sct:derivativeDependency>http://snomed.info/sct/816211006/version/20220131 </sct:derivativeDependency>
        </sct:packageDependency>
    </entry>
```

Does this mean that the `SNOMED CT-International Spanish Edition 2022-10-31 (RF2 SNAPSHOT)` is an extension package that doesn't contain but does depend upon
- the July 2022 International Edition
- and the January 2022 Medical Dictionary for Regulatory Activities simple map module
and their respective transitive dependencies? I.e. the `SNOMED CT-International Spanish Edition 2022-10-31 (RF2 SNAPSHOT)` doesn't contain, but does make references to content from both of these other packages?

That's how I'm understanding it, but I just want to make sure I have it right!

I think the main clarifications we'd need to answer in the documentation for this are
1. How does someone determine whether a URI that is a dependency for their package should be an `editionDependency` or a `derivativeDependency`?
2. What would someone consuming the package do differently for the `editionDependency` URIs compared to the `derivativeDependency` URIs?

I think number 1 could be simply the definition of an extension versus a derivative from the SNOMED CT license? I.e. 
- an extension/edition contains concepts and/or descriptions and/or axioms that aren't in the edition they are extending
- a derivative is essentially just reference set content (map, annotation, subset) or the content of an edition, but does not add concepts, descriptions, or axioms.

Is that what you were thinking?

I think number 2 could be that the consumer then knows if the dependency adds core components, or just reference set content? I'm not sure we'd do anything different with that but perhaps other people might (or future us).

How does that sound?

**Comment by dionmcm** [comment]
Maybe we can suggest a naming convention for edition and extension packages without/with `sct:packageDependency` for example
 - SNOMED CT-AU Edition 30 November 2022 (RF2 SNAPSHOT)
 - SNOMED CT-AU Extension 30 November 2022 (RF2 FULL)

Just to make it clearer for the humans reading it

**Comment by rorydavidson** [comment]
@dionmcm essentially, yes. But this and some other internal discussions are starting to make me lean towards a world where we only have defined editions and not extensions with dependencies to remove the module dependency confusion away from implementers... a discussion for another day!
And agree on the naming convention, although there is no real policing of this once it happens outside of our environments!

**Comment by dionmcm** [comment]
Yep @rorydavidson  I agree on the naming convention too - it is just a convention so there's no way to enforce it, but at least we'd be generally consistent and people that copy us would be similarly consistent. I don't think it is worth shooting for more than that.

We've long had that philosophy on publishing editions vs extensions, the Australian Edition is only published as an edition package, but we'll happily give someone an extension package if they want it...they never have. The logic is that, based on feedback, implementers want a package that's ready to use not the parts and assembly instructions (they have better things to do). To not overburden implementers to even think about the two different options and what it all means (let alone the MDRS) we only publish the edition package which most (so far 100%) want.

On the flip side, those who want the extension package are much more into this sort of thing and are probably very capable of stripping an edition down to an extension if they really needed to. But I suppose having extension packages is good for them too...just for the Australian Edition we've never found anyone who wanted it.

I think the distinction here is an edition being made available as an extension package (i.e. a national extension) versus a derivative which is often an optional extra and naturally fits as an extension rather than edition package.

A practical suggestion to minimise the confusion - if you're going to have edition packages and extension packages maybe put them in two different syndication feeds to prefilter them for implementers and reduce confusion?

The other thing to consider is that RF2 snapshot forms of editions as extension packages don't really make sense. Really an extension package for an edition needs to be a delta or a full that's applied to the underlying edition and then the snapshot calculated using the MDRS (a process that's not that fun or trivial).

In the more simple derivative use case (map, refset, even language translation with additional descriptions) a snapshot extension package being simply appended to a snapshot base edition works and makes more sense. Even extending that to addition of leaf node concepts still works as a very basic extension. But the moment non-leaf concepts or other axioms that affect classification are added that simplistic approach falls down, and the only options are really full or delta extension package or edition package.

In short, snapshot extension packages should be avoided for all but the most trivial editions, and even then for consistency they should probably be avoided.

**Comment by dionmcm** [comment]
So...I might close this long running rambling issue in favour of a few pull requests for these other issues and we can decide them in those PRs. I think the remaining things are

- Naming convention
- Splitting the feeds between extension and edition packages
- Prefer edition packages for editions
- Prefer extension packages for derivatives
- Discourage snapshot extension packages for editions
